### PR TITLE
Feature/mobx

### PR DESCRIPTION
**Pull request by jmoliugp**
todoapp-mobx-demo

### CODE REVIEW

File: src/Scenes/ItemList.js [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
-        onPress={() => props.handleSwitch(props.title)}
-        title="X"
-        color={(props.pending) ? 'green' : 'red'}
-      />
+      <Text style={styles.item}>{props.todoItem.title}</Text>
+      <View style={styles.buttonsView}>
+        <Button
+          onPress={() => props.showEditItem()}
+          title="Edit"
+          color="#841584"
```
Comment (mlvieras) [inline src/Scenes/ItemList.js]: Maybe this color could be set in the component styles

File: src/Scenes/ItemList.js [hunk 2]
```diff
@@ -1,16 +1,25 @@
 import React from 'react';
 import { Text, View, Button } from 'react-native';
+
+import { TodoStore } from '../Stores';
```
Comment (mlvieras) [inline src/Scenes/ItemList.js]: You're not using this import

File: src/Scenes/NewItem.js [hunk 3]
```diff
@@ -65,14 +70,36 @@ export default class NewItem extends Component {
     }
   }
 
+  handleNewTitle = (title) => {
+    const newState = deepcopy(this.state);
+    newState.title = title;
+    this.setState(newState);
```
Comment (mlvieras) [inline src/Scenes/NewItem.js]: You don't need deepcopy here. When setting a state, React automatically shallow merges the old and new states. Since you're only updating the title, you don't need to clone the state.

File: src/Scenes/NewItem.js [hunk 4]
```diff
@@ -65,14 +70,36 @@ export default class NewItem extends Component {
...
+  }
+
+  handleNewPending = (pending) => {
+    const newState = deepcopy(this.state);
+    newState.pending = pending;
+    this.setState(newState);
```
Comment (mlvieras) [inline src/Scenes/NewItem.js]: Here too

File: src/Scenes/TodoList.js [hunk 5]
```diff
@@ -45,15 +42,12 @@ export default class TodoList extends Component {
 
   constructor(props) {
     super(props);
-    const pendingItems = this.newItems(preloadItems.pendingItems, true);
-    const doneItems = this.newItems(preloadItems.doneItems, false);
     this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent);
     this.props.navigator.setButtons(tabButtons);
```
Comment (mcousillas6) [inline src/Scenes/TodoList.js]: There is a static `navigatorButtons` props that you can use on your components to get this behavior, similar to `navigatorStyle`.

File: src/Scenes/TodoList.js [hunk 6]
```diff
@@ -100,23 +96,31 @@ export default class TodoList extends Component {
   renderItem = ({ item }) => {
     return (
       <ItemList
-        title={item.title}
-        pending={item.pending}
-        handleSwitch={this.handleSwitch}
+        todoItem={item}
+        handleSwitch={() => {
+          const newItem = Object.assign({}, item);
```
Comment (mlvieras) [inline src/Scenes/TodoList.js]: You can also use `const newItem = {...item}`.

File: src/Scenes/TodoList.js [hunk 7]
```diff
@@ -69,28 +63,30 @@ export default class TodoList extends Component {
...
-  newItems = (items, pending) => {
-    return items.map((item) => {
-      return { title: item, pending };
+  showEditItem = () => {
+    this.props.navigator.showModal({
+      screen: 'TodoList.NewItem',
+      title: 'Edit Item',
+      passProps: {
+        isEditable: true,
+        handleNewItem: this.handleNewItem,
```
Comment (mcousillas6) [inline src/Scenes/TodoList.js]: I think you don't need this callback anymore, since you are calling `TodoStore.edit/createItem` inside the NewItem component.

File: src/Scenes/TodoList.js [hunk 8]
```diff
@@ -69,28 +63,30 @@ export default class TodoList extends Component {
...
 
-  newItems = (items, pending) => {
-    return items.map((item) => {
-      return { title: item, pending };
+  showEditItem = () => {
+    this.props.navigator.showModal({
+      screen: 'TodoList.NewItem',
+      title: 'Edit Item',
+      passProps: {
+        isEditable: true,
```
Comment (mcousillas6) [inline src/Scenes/TodoList.js]: Same here, you don't need a flag for this I think, you can just check if there is a `TodoStore.selectedTodo` set or not.

File: src/Stores/TodoStore.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import { observable, computed, action } from 'mobx';
+
+const pendingItems = ['Cofee', 'Fruit'];
+const doneItems = ['Monitors', 'Notebooks', 'PCs'];
+
+const genTodoId = () => Math.random();
```
Comment (mlvieras) [inline src/Stores/TodoStore.js]: Please don't abbreviate function names, `generateTodoId` is more clear and you don't span too much space.

File: src/Stores/TodoStore.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+const genTodoList = (items, pending) => {
+  return items.map((item) => {
+    return { title: item, pending, id: genTodoId() };
+  });
+};
+
+const preloadItems = genTodoList(pendingItems, true).concat(genTodoList(doneItems, false));
+
+class TodoStore {
+  debugger;
```
Comment (mlvieras) [inline src/Stores/TodoStore.js]: Many debugger calls on this file

File: src/Scenes/NewItem.js [hunk 11]
```diff
@@ -41,21 +46,21 @@ const navButtons = {
...
+@observer
 export default class NewItem extends Component {
   constructor(props) {
     super(props);
     this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent.bind(this));
     this.props.navigator.setButtons(navButtons);
-    this.state = {
-      title: '',
-    };
+    this.state = (this.props.isEditable) ? TodoStore.selectedTodo : { title: '', pending: true, id: genTodoId() };
```
Comment (mcousillas6) [inline src/Scenes/NewItem.js]: You can remove the `isEditable` flag and do something like `TodoStore.selectedTodo || { title: '', pending: true, id: genTodoId() }

Comment (mcousillas6) [reply src/Scenes/NewItem.js]: Also, avoid using `''` as empty strings values, you should use `null` so placeholders are rendered on `Text` components.

File: src/Stores/TodoStore.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @action removeTodo = (item) => {
+    this.todos = this.todos.filter(i => i.id !== item.id);
+  }
+
+  @action editTodo = (item) => {
+    debugger;
+    this.todos = this.todos.map((i) => {
+      return (i.id === item.id) ? item : i;
```
Comment (mlvieras) [inline src/Stores/TodoStore.js]: You can write this as an arrow function with an implicit return:
```javascript
this.todos = this.todos.map((i) => (i.id === item.id) ? item : i);
```

File: src/Scenes/NewItem.js [hunk 13]
```diff
@@ -65,14 +70,36 @@ export default class NewItem extends Component {
     }
   }
 
+  handleNewTitle = (title) => {
+    const newState = deepcopy(this.state);
```
Comment (mcousillas6) [inline src/Scenes/NewItem.js]: Why are you using deepcopy for this? a simple `this.setState({...this.state, title: title})` would work.

File: src/Stores/TodoStore.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    this.selectedTodo = item;
+  }
+
+  @action unSelectTodo = () => {
+    this.selectedTodo = null;
+  }
+}
+
+export const Store = new TodoStore();
+export const genId = genTodoId;
```
Comment (mlvieras) [inline src/Stores/TodoStore.js]: Maybe I'm confused, but you're exporting `genTodoId` named as `genId`. Yet you always import it as `genTodoId`. Is this working OK?

File: src/Scenes/NewItem.js (hunk 11 above)
Comment (mlvieras) [inline src/Scenes/NewItem.js]: I think you tried to merge a component's state with a Mobx Observable. I don't think this is the best approach. When using Mobx, avoid assigning an observable to a component's state. Instead, use Mobx actions to update the observable, observe what you need and Mobx will update the components for you.

We can talk about this if you want, maybe I didn't understand what you were trying to accomplish

Comment (mcousillas6) [reply src/Scenes/NewItem.js]: Yes, if you really need to store something mobx related on the state, use the `toJS` function that mobx provides to transform it.

File: src/Scenes/TodoList.js [hunk 15]
```diff
@@ -69,28 +63,30 @@ export default class TodoList extends Component {
...
 
-  handleSwitch = (title) => {
-    const itemsUpdated = this.state.items.map((item) => {
-      const pending = (item.title === title) ? !item.pending : item.pending;
-      return { title: item.title, pending };
-    });
-    this.setState({ items: itemsUpdated });
+  //Store Handlers
+
+  handleSwitch = (item) => {
```
Comment (santiagofm) [inline src/Scenes/TodoList.js]: You don't need this handler anymore also

File: src/Stores/TodoStore.js (hunk 14 above)
Comment (santiagofm) [reply src/Stores/TodoStore.js]: Yes but he imports it and re exports it as `genTodoId` in the `index.js`

Comment (mlvieras) [reply src/Stores/TodoStore.js]: Missed that completely, thanks!

File: src/Scenes/ItemList.js [hunk 16]
```diff
@@ -12,9 +11,9 @@ const ItemList = (props) => {
         <Button
           onPress={() => props.showEditItem()}
           title="Edit"
-          color="#841584"
         />
         <Button
+          style={styles.editButton}
```
Comment (mlvieras) [inline src/Scenes/ItemList.js]: Why did you move this style from one button to the other?

File: src/Scenes/TodoList.js (hunk 5 above)
Comment (jmoliugp) [reply src/Scenes/TodoList.js]: Actually in this case I am not able to declare the navigatorButtons as a static variable because this component is being wrapped in a HOC.
